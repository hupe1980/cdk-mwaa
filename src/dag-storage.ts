import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import { Construct } from 'constructs';

/**
 * Configuration options for S3 path storage.
 */
export interface DagStorageConfigOptions {
  /** The S3 path where the resource is stored */
  readonly s3Path: string;
  /** Optional object version for versioned buckets */
  readonly s3ObjectVersion?: string;
}

/**
 * Options for deploying files to the DAG storage bucket.
 */
export interface DagStorageDeployOptions {
  /** Sources to deploy into the bucket */
  readonly sources?: s3deploy.ISource[];
  /** Whether to retain files on deletion of the stack */
  readonly retainOnDelete?: boolean;
  /** Whether to prune old versions of deployed files */
  readonly prune?: boolean;
}

/**
 * Properties for configuring the DAG storage bucket.
 */
export interface DagStorageProps {
  /** Optional custom bucket name */
  readonly bucketName?: string;
  /** Enable versioning for the bucket */
  readonly versioned?: boolean;
  /** Lifecycle rule for expiring non-current versions */
  readonly noncurrentVersionExpiration?: cdk.Duration;
  /** Policy for bucket removal */
  readonly removalPolicy?: cdk.RemovalPolicy;
  /** Path for storing DAG files */
  readonly dagS3Path?: string;
  /** Configuration for plugins storage */
  readonly pluginsConfig?: DagStorageConfigOptions;
  /** Configuration for requirements storage */
  readonly requirementsConfig?: DagStorageConfigOptions;
  /** Configuration for startup script storage */
  readonly startupScriptConfig?: DagStorageConfigOptions;
  /** Options for deploying files into the bucket */
  readonly deployOptions?: DagStorageDeployOptions;
}

/**
 * Represents a storage location for MWAA DAGs and dependencies in an S3 bucket.
 */
export class DagStorage extends Construct {
  /** The S3 bucket storing DAGs, plugins, requirements, and startup scripts */
  public readonly bucket: s3.Bucket;
  /** S3 path for DAGs */
  public readonly dagS3Path?: string;
  /** Plugin storage configuration */
  public readonly pluginsConfig?: DagStorageConfigOptions;
  /** Requirements storage configuration */
  public readonly requirementsConfig?: DagStorageConfigOptions;
  /** Startup script storage configuration */
  public readonly startupScriptConfig?: DagStorageConfigOptions;

  constructor(scope: Construct, id: string, props: DagStorageProps) {
    super(scope, id);

    // Create S3 bucket with defined properties
    this.bucket = new s3.Bucket(this, 'DagStorageBucket', {
      bucketName: props.bucketName,
      versioned: props.versioned,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: props.removalPolicy ?? cdk.RemovalPolicy.RETAIN,
      autoDeleteObjects: props.removalPolicy === cdk.RemovalPolicy.DESTROY,
    });

    // Apply lifecycle rule if versioning and expiration settings are provided
    if (props.versioned && props.noncurrentVersionExpiration) {
      this.bucket.addLifecycleRule({
        id: 'DeleteOldVersions',
        noncurrentVersionExpiration: props.noncurrentVersionExpiration,
      });
    }

    // Assign configuration properties
    this.dagS3Path = props.dagS3Path ?? 'dags/';
    this.pluginsConfig = props.pluginsConfig;
    this.requirementsConfig = props.requirementsConfig;
    this.startupScriptConfig = props.startupScriptConfig;

    // Deploy specified sources to the bucket if provided
    if (props.deployOptions?.sources && props.deployOptions?.sources.length > 0) {
      new s3deploy.BucketDeployment(this, 'DagStorageDeployment', {
        sources: props.deployOptions.sources,
        destinationBucket: this.bucket,
        retainOnDelete: props.deployOptions.retainOnDelete ?? false,
        prune: props.deployOptions.prune ?? true,
      });
    }
  }
}
