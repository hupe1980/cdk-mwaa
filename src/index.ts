export * from './dag-storage';
export * from './sizing';
export * from './environment';
export * from './patterns';
export * from './security-group';