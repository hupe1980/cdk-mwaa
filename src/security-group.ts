import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

/**
 * Properties for defining a Security Group.
 */
export interface SecurityGroupProps extends ec2.SecurityGroupProps {}

/**
 * A custom Security Group with a self-referencing ingress rule for MWAA.
 */
export class SecurityGroup extends ec2.SecurityGroup {
  /**
   * Creates a new Security Group with self-referencing ingress rules.
   *
   * @param scope The parent construct.
   * @param id The unique identifier for this construct.
   * @param props Security group properties.
   */
  constructor(scope: Construct, id: string, props: SecurityGroupProps) {
    super(scope, id, props);

    // Allow all inbound traffic within the security group
    this.addIngressRule(this, ec2.Port.allTraffic(), 'MWAA Self-referencing rule');
  }
}
