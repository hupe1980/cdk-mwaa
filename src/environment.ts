import { Lazy, Stack, Tags } from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as mwaa from 'aws-cdk-lib/aws-mwaa';
import { Construct } from 'constructs';

import { DagStorage } from './dag-storage';
import { SecurityGroup } from './security-group';
import { Sizing } from './sizing';

/**
 * Enum for the webserver access mode of the MWAA environment.
 */
export enum WebserverAccessMode {
  PRIVATE_ONLY = 'PRIVATE_ONLY',
  PUBLIC_ONLY = 'PUBLIC_ONLY',
}

/**
 * Enum for the endpoint management type for the MWAA environment.
 */
export enum EndpointManagement {
  CUSTOMER = 'CUSTOMER',
  SERVICE = 'SERVICE',
}

/**
 * Enum for the log level for Apache Airflow.
 */
export enum LogLevel {
  CRITICAL = 'CRITICAL',
  ERROR = 'ERROR',
  WARNING = 'WARNING',
  INFO = 'INFO',
  DEBUG = 'DEBUG',
}

/**
 * Options for configuring the Secrets backend.
 */
export interface SecretsBackendOptions {
  readonly connectionsPrefix?: string;
  readonly connectionsLookupPattern?: string;
  readonly variablesPrefix?: string;
  readonly variablesLookupPattern?: string;
}

/**
 * Options for configuring the Email backend.
 */
export interface EmailBackendOptions {
  readonly fromEmail: string;
  readonly connId?: string;
}

/**
 * Defines the logging configuration properties for various Airflow log types.
 */
export interface LoggingConfigurationProperty {
  /**
   * Indicates whether to enable the Apache Airflow log type (e.g. DagProcessingLogs) in CloudWatch Logs.
   */
  readonly enabled?: boolean;

  /**
   * Defines the log level for the specified log type (e.g. DagProcessingLogs).
   * Valid values: CRITICAL, ERROR, WARNING, INFO, DEBUG.
   */
  readonly logLevel?: LogLevel;
}

/**
 * Logging configuration for the MWAA environment.
 */
export interface LoggingConfiguration {
  readonly dagProcessingLogs?: LoggingConfigurationProperty;
  readonly schedulerLogs?: LoggingConfigurationProperty;
  readonly taskLogs?: LoggingConfigurationProperty;
  readonly webserverLogs?: LoggingConfigurationProperty;
  readonly workerLogs?: LoggingConfigurationProperty;
}

/**
 * Properties for creating an MWAA environment.
 */
export interface EnvironmentProps {
  readonly airflowVersion: string; // Version of Apache Airflow to use
  readonly airflowConfigurationOptions?: { [key: string]: any }; // Airflow configuration options
  readonly name: string; // The name of the MWAA environment
  readonly dagStorage: DagStorage; // Storage for the DAGs
  readonly vpc: ec2.IVpc; // VPC where MWAA environment is created
  readonly securityGroups?: ec2.ISecurityGroup[]; // Security groups for the MWAA environment
  readonly webserverAccessMode?: WebserverAccessMode; // Webserver access mode (default: PUBLIC_ONLY)
  readonly endpointManagement?: EndpointManagement; // Defines whether the VPC endpoints are managed by the customer or by MWAA
  readonly sizing: Sizing; // Sizing configuration for MWAA environment
  readonly kmsKey?: kms.IKey; // KMS key for encryption (optional)
  readonly loggingConfiguration?: LoggingConfiguration; // Airflow logs configuration
  readonly weeklyMaintenanceWindowStart?: string; // Weekly maintenance window start time
  readonly tags?: { [key: string]: string }; // Tags to apply to MWAA environment
}

/**
 * Represents an MWAA environment.
 */
export class Environment extends Construct {
  public readonly name: string; // Name of the MWAA environment
  public readonly arn: string; // ARN of the MWAA environment
  public readonly celeryExecutorQueue: string; // ARN for Celery Executor queue
  public readonly databaseVpcEndpointService: string; // VPC endpoint for RDS database
  public readonly dagProcessingLogsGroupArn: string; // ARN for DAG processing logs group
  public readonly schedulerLogsGroupArn: string; // ARN for scheduler logs group
  public readonly taskLogsGroupArn: string; // ARN for task logs group
  public readonly webserverLogsGroupArn: string; // ARN for webserver logs group
  public readonly workerLogsGroupArn: string; // ARN for worker logs group
  public readonly webserverUrl: string; // URL for the web server
  public readonly webserverVpcEndpointService: string; // VPC endpoint for web server

  private airflowConfigurationOptions: { [key: string]: any }; // Airflow configuration options
  private executionRole: iam.Role; // IAM execution role for MWAA environment

  /**
   * Creates an MWAA environment.
   * @param scope - The scope of the construct.
   * @param id - The unique ID of the construct.
   * @param props - The properties for the MWAA environment.
   */
  constructor(scope: Construct, id: string, props: EnvironmentProps) {
    super(scope, id);

    const { region, account } = Stack.of(this);

    // Default Airflow configuration options
    this.airflowConfigurationOptions = {
      'scheduler.catchup_by_default': false,
      ...props.airflowConfigurationOptions ?? {},
    };

    const executionPolicy = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['airflow:PublishMetrics'],
          resources: [`arn:aws:airflow:${region}:${account}:environment/${props.name}`],
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.DENY,
          actions: ['s3:ListAllMyBuckets'],
          resources: ['*'],
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['s3:GetObject*', 's3:GetBucket*', 's3:List*'],
          resources: [props.dagStorage.bucket.bucketArn, `${props.dagStorage.bucket.bucketArn}/*`],
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            'logs:CreateLogStream',
            'logs:CreateLogGroup',
            'logs:PutLogEvents',
            'logs:GetLogEvents',
            'logs:GetLogRecord',
            'logs:GetLogGroupFields',
            'logs:GetQueryResults',
          ],
          resources: [
            `arn:aws:logs:${region}:${account}:log-group:airflow-${props.name}-*`,
          ],
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['cloudwatch:PutMetricData'],
          resources: ['*'],
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['sqs:ChangeMessageVisibility', 'sqs:DeleteMessage', 'sqs:GetQueueAttributes', 'sqs:ReceiveMessage'],
          resources: [`arn:aws:sqs:${region}:*:airflow-celery-*`],
        }),
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            'kms:Decrypt',
            'kms:DescribeKey',
            'kms:GenerateDataKey*',
            'kms:Encrypt',
          ],
          notResources: [`arn:aws:kms:*:${account}:key/*`],
          conditions: {
            StringLike: {
              'kms:ViaService': [`sqs.${region}.amazonaws.com`],
            },
          },
        }),
      ],
    });

    // Cross-service confused deputy prevention
    const conditions = {
      ArnLike: {
        'aws:SourceArn': `arn:aws:airflow:${region}:${account}:environment/${props.name}`,
      },
      StringEquals: {
        'aws:SourceAccount': account,
      },
    };

    // Creating the IAM execution role
    this.executionRole = new iam.Role(this, 'MWAAExecutionRole', {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal('airflow.amazonaws.com', { conditions }),
        new iam.ServicePrincipal('airflow-env.amazonaws.com', { conditions }),
      ),
      path: '/service-role/',
      inlinePolicies: {
        mwaaExecutionPolicy: executionPolicy,
      },
    });

    const weeklyMaintenanceWindowStart = this.validateWeeklyMaintenanceWindowStart(props.weeklyMaintenanceWindowStart);

    // Security group for MWAA environment
    const securityGroups = props.securityGroups ?? [new SecurityGroup(this, 'MWAAEnvironmentSecurityGroup', {
      vpc: props.vpc,
    })];

    // Creating the MWAA environment
    const environment = new mwaa.CfnEnvironment(this, 'MWAAEnvironment', {
      airflowVersion: props.airflowVersion,
      airflowConfigurationOptions: Lazy.any({ produce: () => this.airflowConfigurationOptions }),
      name: props.name,
      executionRoleArn: this.executionRole.roleArn,
      sourceBucketArn: props.dagStorage.bucket.bucketArn,
      dagS3Path: props.dagStorage.dagS3Path,
      pluginsS3Path: props.dagStorage.pluginsS3Path,
      pluginsS3ObjectVersion: props.dagStorage.pluginsS3ObjectVersion,
      requirementsS3Path: props.dagStorage.requirementsS3Path,
      requirementsS3ObjectVersion: props.dagStorage.requirementsS3ObjectVersion,
      startupScriptS3Path: props.dagStorage.startupScriptS3Path,
      startupScriptS3ObjectVersion: props.dagStorage.startupScriptS3ObjectVersion,
      environmentClass: props.sizing.environmentClass,
      maxWorkers: props.sizing.maxWorkers,
      minWorkers: props.sizing.minWorkers,
      maxWebservers: props.sizing.maxWebservers,
      minWebservers: props.sizing.minWebservers,
      schedulers: props.sizing.schedulers,
      webserverAccessMode: props.webserverAccessMode ?? WebserverAccessMode.PUBLIC_ONLY,
      networkConfiguration: {
        securityGroupIds: securityGroups.map(sg => sg.securityGroupId),
        subnetIds: props.vpc.privateSubnets.map(subnet => subnet.subnetId),
      },
      endpointManagement: props.endpointManagement,
      loggingConfiguration: props.loggingConfiguration,
      weeklyMaintenanceWindowStart: weeklyMaintenanceWindowStart,
    });

    environment.node.addDependency(props.dagStorage);
    environment.node.addDependency(this.executionRole);

    // Assign environment properties
    this.name = props.name;
    this.arn = environment.attrArn;
    this.celeryExecutorQueue = environment.attrCeleryExecutorQueue;
    this.databaseVpcEndpointService = environment.attrDatabaseVpcEndpointService;
    this.dagProcessingLogsGroupArn = environment.attrLoggingConfigurationDagProcessingLogsCloudWatchLogGroupArn;
    this.schedulerLogsGroupArn = environment.attrLoggingConfigurationSchedulerLogsCloudWatchLogGroupArn;
    this.taskLogsGroupArn = environment.attrLoggingConfigurationTaskLogsCloudWatchLogGroupArn;
    this.webserverLogsGroupArn = environment.attrLoggingConfigurationWebserverLogsCloudWatchLogGroupArn;
    this.workerLogsGroupArn = environment.attrLoggingConfigurationWorkerLogsCloudWatchLogGroupArn;
    this.webserverUrl = environment.attrWebserverUrl;
    this.webserverVpcEndpointService = environment.attrWebserverVpcEndpointService;

    // Apply tags if any
    if (props.tags) {
      for (const [key, value] of Object.entries(props.tags)) {
        Tags.of(environment).add(key, value);
      }
    }
  }

  /**
   * Enables the use of AWS Secrets Manager as a backend for storing Airflow connections and variables.
   * @param options - The configuration options for Secrets Manager backend.
   */
  public enableSecretsBackend(options?: SecretsBackendOptions): void {
    const defaultOptions = {
      connectionsPrefix: 'airflow/connections',
      variablesPrefix: 'airflow/variables',
    };

    const finalOptions = {
      ...defaultOptions,
      ...options,
    };

    this.airflowConfigurationOptions['secrets.backend'] = 'airflow.providers.amazon.aws.secrets.secrets_manager.SecretsManagerBackend';
    this.airflowConfigurationOptions['secrets.backend_kwargs'] = JSON.stringify({
      connections_prefix: finalOptions.connectionsPrefix,
      connections_lookup_pattern: finalOptions.connectionsLookupPattern,
      variables_prefix: finalOptions.variablesPrefix,
      variables_lookup_pattern: finalOptions.variablesLookupPattern,
    });

    const { region, account } = Stack.of(this);

    // Adding permissions for the Secrets Manager backend
    this.addToRolePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: [
        'secretsmanager:GetResourcePolicy',
        'secretsmanager:GetSecretValue',
        'secretsmanager:DescribeSecret',
        'secretsmanager:ListSecretVersionIds',
      ],
      resources: [
        `arn:aws:secretsmanager:${region}:${account}:secret:${finalOptions.connectionsPrefix}/*`,
        `arn:aws:secretsmanager:${region}:${account}:secret:${finalOptions.variablesPrefix}/*`,
      ],
      conditions: {
        StringEquals: {
          'aws:ResourceTag/Environment': 'MWAA',
        },
      },
    }));

    this.addToRolePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['secretsmanager:ListSecrets'],
      resources: ['*'],
      conditions: {
        StringEquals: {
          'aws:ResourceTag/Environment': 'MWAA',
        },
      },
    }));
  }

  /**
   * Enables the email backend for Airflow to send email notifications.
   * @param options - The configuration options for the email backend.
   */
  public enableEmailBackend(options: EmailBackendOptions): void {
    const connId = options.connId ?? 'aws_default';

    this.airflowConfigurationOptions['email.email_backend'] = 'airflow.providers.amazon.aws.utils.emailer.send_email';
    this.airflowConfigurationOptions['email.from_email'] = options.fromEmail;
    this.airflowConfigurationOptions['email.conn_id'] = connId;

    if (connId === 'aws_default') {
      this.addToRolePolicy(new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['ses:SendEmail'],
        resources: ['*'],
      }));
    }
  }

  /**
   * Adds a policy statement to the execution role's policy.
   *
   * @param statement - The IAM policy statement to add to the role's policy.
   */
  public addToRolePolicy(statement: iam.PolicyStatement): void {
    this.executionRole.addToPolicy(statement);
  }

  /**
   * Sets an Airflow configuration option.
   *
   * @param key - The configuration option key.
   * @param value - The configuration option value.
   * @returns void
   */
  public setAirflowConfigurationOption(key: string, value: any): void {
    this.airflowConfigurationOptions[key] = value;
  }

  /**
   * Validates the weekly maintenance window start time.
   * @param weeklyMaintenanceWindowStart - The weekly maintenance window start time in UTC (e.g., 'Sun:10:00').
   */
  private validateWeeklyMaintenanceWindowStart(weeklyMaintenanceWindowStart?: string): string | undefined {
    if (weeklyMaintenanceWindowStart) {
      const regex = /^([a-zA-Z]+):([0-9]{2}):([0-9]{2})$/;
      if (!weeklyMaintenanceWindowStart.match(regex)) {
        throw new Error('Invalid format for weeklyMaintenanceWindowStart. Expected format: "Day:HH:mm"');
      }
      return weeklyMaintenanceWindowStart;
    }
    return undefined;
  }
}
