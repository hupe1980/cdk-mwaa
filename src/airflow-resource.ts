import * as path from 'node:path';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as customResources from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { Environment } from './environment';


export interface AirflowResourceBaseProps {
  readonly environment: Environment;
  readonly resourceType: string;
  readonly properties: Record<string, any>;
  readonly airflowRole?: string;
}


export abstract class AirflowResourceBase extends Construct {
  constructor(scope: Construct, id: string, props: AirflowResourceBaseProps) {
    super(scope, id);

    const { region, account } = cdk.Stack.of(this);
    const mwaaEnvName = props.environment.name;
    const airlfowRole = props.airflowRole ?? 'Op';

    // IAM Role for Lambda
    const lambdaRole = new iam.Role(this, 'Role', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    });

    lambdaRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
    );

    lambdaRole.addToPolicy(
      new iam.PolicyStatement({
        actions: ['airflow:InvokeRestApi'],
        resources: [`arn:aws:airflow:${region}:${account}:role/${mwaaEnvName}/${airlfowRole}`],
      }),
    );

    // Lambda Function
    const airflowLambda = new lambda.SingletonFunction(this, 'Lambda', {
      uuid: `github.com/hupe1980/cdk-mwaa/airflow/${props.resourceType}`,
      runtime: lambda.Runtime.PYTHON_3_11,
      handler: `${props.resourceType}.on_event`,
      code: lambda.Code.fromAsset(path.join(__dirname, 'lambdas')),
      timeout: cdk.Duration.seconds(30),
      role: lambdaRole,
    });

    // Custom Resource Provider
    const provider = new customResources.Provider(this, 'Provider', {
      onEventHandler: airflowLambda,
      logGroup: new logs.LogGroup(this, 'ProviderLogGroup', {
        retention: logs.RetentionDays.ONE_DAY,
      }),
    });

    // Custom Resource to trigger Lambda
    new cdk.CustomResource(this, 'Resource', {
      serviceToken: provider.serviceToken,
      properties: {
        MwaaEnvName: mwaaEnvName,
        ...props.properties,
      },
    });

    // Add dependency on the MWAA environment
    this.node.addDependency(props.environment);
  }
}
