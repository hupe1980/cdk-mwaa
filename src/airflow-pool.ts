import { Construct } from 'constructs';
import { AirflowResourceBase } from './airflow-resource';
import { Environment } from './environment';

export interface AirflowPoolProps {
  readonly environment: Environment;
  readonly poolName: string;
  readonly poolSlots?: number;
  readonly poolDescription?: string;
}

export class AirflowPool extends AirflowResourceBase {
  constructor(scope: Construct, id: string, props: AirflowPoolProps) {
    super(scope, id, {
      environment: props.environment,
      resourceType: 'pool',
      properties: {
        PoolName: props.poolName,
        PoolSlots: props.poolSlots ?? 3,
        PoolDescription: props.poolDescription ?? '',
      },
    });
  }
}
