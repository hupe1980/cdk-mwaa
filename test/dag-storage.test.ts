import * as cdk from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { DagStorage, DagStorageProps } from '../src';

describe('DagStorage Snapshot Tests', () => {
  let app: cdk.App;
  let stack: cdk.Stack;

  beforeEach(() => {
    app = new cdk.App();
    stack = new cdk.Stack(app, 'TestStack');

    const props: DagStorageProps = {
      bucketName: 'test-bucket',
      versioned: true,
      noncurrentVersionExpiration: cdk.Duration.days(30),
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      dagsOptions: { s3Path: 'dags/' },
      configsOptions: {
        s3Prefix: 'configs/',
        requirements: { name: 'requirements.txt' },
        plugins: { name: 'plugins.zip' },
        startupScript: { name: 'startup.sh' },
      },
    };

    new DagStorage(stack, 'DagStorage', props);
  });

  test('matches snapshot', () => {
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  test('S3 Bucket has correct properties', () => {
    const template = Template.fromStack(stack);
    template.hasResourceProperties('AWS::S3::Bucket', {
      BucketName: 'test-bucket',
      VersioningConfiguration: { Status: 'Enabled' },
      LifecycleConfiguration: Match.objectLike({
        Rules: [
          {
            NoncurrentVersionExpiration: { NoncurrentDays: 30 },
            Status: 'Enabled',
          },
        ],
      }),
    });
  });
});