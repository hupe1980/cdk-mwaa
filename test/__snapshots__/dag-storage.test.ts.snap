// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DagStorage Snapshot Tests matches snapshot 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs20.x",
      },
      "cn-northwest-1": {
        "value": "nodejs20.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs20.x",
      },
      "us-gov-west-1": {
        "value": "nodejs20.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "faa95a81ae7d7373f3e1f242268f904eb748d8d0fdd306e8a6fe515a1905a7d6.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "DagStorageDagStorageBucket1A3B53A2",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DagStorageDagStorageBucket1A3B53A2": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "test-bucket",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "Id": "DeleteOldVersions",
              "NoncurrentVersionExpiration": {
                "NoncurrentDays": 30,
              },
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "DagStorageDagStorageBucketAutoDeleteObjectsCustomResource0BD31144": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "DagStorageDagStorageBucketPolicy250EE88D",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "DagStorageDagStorageBucket1A3B53A2",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "DagStorageDagStorageBucketPolicy250EE88D": {
      "Properties": {
        "Bucket": {
          "Ref": "DagStorageDagStorageBucket1A3B53A2",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "DagStorageDagStorageBucket1A3B53A2",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "DagStorageDagStorageBucket1A3B53A2",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
